File: owner_net.js
window.COMMONS_OWNER_NET = "hashed-ip-door";
(function () {
  // Directive 10. Cite BRYCE-1787134106972-vr8fo8. Do not remint.
  // Law: admin-no-verification-loop-20260819-01. Do not remint.
  // GitHub Pages is static. The browser hashes this network's public IP
  // (pepper + LF + IP) and never writes the address. A match against an
  // enrolled slot annotates visible from=BRYCE as optional context. No login.
  // It cannot control participation, reads, writes, or execution. from= stays
  // a claim.
  // Two slots: pc and phone. Same public IP is not the door.
  // Phone on cell and PC at home must persist two different digests.
  // A remembered-BRYCE browser publishes {k,sha256,via}. Matching does
  // not use the ntfy bus (that is the same-NAT toy). Persist is owner_net.py.
  if (window.COMMONS_OWNER_BOOT) return;
  window.COMMONS_OWNER_BOOT = 1;

  var SPEC_NAME = "owner.json";
  var ONLY_CLAIM = "BRYCE";
  var FALLBACK_PEPPER = "commons-owner-v1";
  var HASH_RE = /^[0-9a-f]{64}$/;
  var NET_TOPIC = "woahwhattheheck-commons-owner-net";
  var NET_HOSTS = [
    "https://ntfy.sh",
    "https://ntfy.envs.net",
    "https://ntfy.adminforge.de",
    "https://ntfy.mzte.de"
  ];
  var SEND_GAP_MS = 6 * 60 * 60 * 1000;
  var SENT_KEY = "commons-owner-net-sent";

  function assetUrl(name) {
    var link = document.querySelector('link[rel="stylesheet"]');
    var href = (link && link.getAttribute("href")) || "./commons.css";
    return href.replace(/commons\.css.*$/, name);
  }

  function normalizeIp(raw) {
    var s = String(raw || "").trim();
    if (s.charAt(0) === "[" && s.charAt(s.length - 1) === "]") s = s.slice(1, -1);
    var pct = s.indexOf("%");
    if (pct !== -1) s = s.slice(0, pct);
    if (s.indexOf(":") !== -1) s = s.toLowerCase();
    return s;
  }

  function looksLikeIp(s) {
    s = String(s || "");
    if (/^\d{1,3}(\.\d{1,3}){3}$/.test(s)) return true;
    if (s.indexOf(":") !== -1 && /^[0-9a-f:]+$/.test(s)) return true;
    return false;
  }

  function asHash(item) {
    var h = "";
    if (typeof item === "string") h = item.toLowerCase();
    else if (item && typeof item.sha256 === "string") h = item.sha256.toLowerCase();
    return HASH_RE.test(h) ? h : "";
  }

  function sha256hex(s) {
    if (!window.crypto || !crypto.subtle || typeof TextEncoder === "undefined") {
      return Promise.reject(new Error("no-subtle"));
    }
    var bytes = new TextEncoder().encode(s);
    return crypto.subtle.digest("SHA-256", bytes).then(function (buf) {
      var b = new Uint8Array(buf);
      var out = "";
      var i;
      for (i = 0; i < b.length; i++) {
        out += (b[i] < 16 ? "0" : "") + b[i].toString(16);
      }
      return out;
    });
  }

  function digestIp(ip, pepper) {
    var n = normalizeIp(ip);
    if (!looksLikeIp(n)) return Promise.reject(new Error("not-ip"));
    return sha256hex(String(pepper || FALLBACK_PEPPER) + "\n" + n);
  }

  function fetchText(url) {
    return fetch(url, { cache: "no-store", credentials: "omit" }).then(function (r) {
      if (!r.ok) throw new Error("echo");
      return r.text();
    });
  }

  function hashEcho(url, pepper) {
    return fetchText(url).then(function (text) {
      return digestIp(text, pepper);
    });
  }

  function loadSpec() {
    return fetch(assetUrl(SPEC_NAME) + "?v=" + Date.now(), {
      cache: "no-store",
      credentials: "omit"
    }).then(function (r) {
      if (!r.ok) throw new Error("spec");
      return r.json();
    }, function () {
      return {
        claim: ONLY_CLAIM,
        algo: "sha256",
        pepper: FALLBACK_PEPPER,
        slots: { pc: null, phone: null },
        hashes: []
      };
    });
  }

  function echoList(spec) {
    var echoes = (spec && spec.echoes) || {};
    var list = echoes.browser || echoes.host || [];
    if (!list.length) {
      list = ["https://api.ipify.org", "https://api64.ipify.org"];
    }
    var out = [];
    var seen = {};
    list.forEach(function (u) {
      u = String(u || "").trim();
      if (!u || seen[u] || u.indexOf("https://") !== 0) return;
      seen[u] = 1;
      out.push(u);
    });
    return out;
  }

  function collectDigests(spec) {
    var pepper = String((spec && spec.pepper) || FALLBACK_PEPPER);
    var urls = echoList(spec);
    return Promise.all(urls.map(function (url) {
      return hashEcho(url, pepper).then(function (h) { return h; }, function () { return ""; });
    })).then(function (got) {
      var uniq = [];
      var seen = {};
      got.forEach(function (h) {
        if (!HASH_RE.test(h) || seen[h]) return;
        seen[h] = 1;
        uniq.push(h);
      });
      return uniq;
    });
  }

  function slotHash(spec, via) {
    var slots = (spec && spec.slots) || {};
    return asHash(slots[via]);
  }

  function slotsDistinct(spec) {
    var pc = slotHash(spec, "pc");
    var phone = slotHash(spec, "phone");
    return !!(pc && phone && pc !== phone);
  }

  function deviceVia() {
    try {
      if (navigator.userAgentData && navigator.userAgentData.mobile) return "phone";
    } catch (e) {}
    var ua = String(navigator.userAgent || "");
    if (/Android|iPhone|iPod|webOS|BlackBerry|IEMobile|Opera Mini/i.test(ua)) return "phone";
    return "pc";
  }

  function rememberedBryce() {
    try {
      return String(localStorage.getItem("commons-from") || "").toUpperCase() === ONLY_CLAIM;
    } catch (e) {
      return false;
    }
  }

  function recentlySent(digest, via) {
    try {
      var raw = localStorage.getItem(SENT_KEY);
      if (!raw) return false;
      var o = JSON.parse(raw);
      if (!o || o.hash !== digest || o.via !== via) return false;
      return (Date.now() - Number(o.ts || 0)) < SEND_GAP_MS;
    } catch (e) {
      return false;
    }
  }

  function markSent(digest, via) {
    try {
      localStorage.setItem(SENT_KEY, JSON.stringify({ hash: digest, via: via, ts: Date.now() }));
    } catch (e) {}
  }

  function publishDigest(digest) {
    if (!HASH_RE.test(digest)) return Promise.resolve(false);
    var via = deviceVia();
    if (recentlySent(digest, via)) return Promise.resolve(false);
    var packed = JSON.stringify({ k: "owner-net", sha256: digest, via: via });
    function send(i) {
      if (i >= NET_HOSTS.length) return Promise.resolve(false);
      return fetch(NET_HOSTS[i] + "/" + NET_TOPIC, {
        method: "POST",
        credentials: "omit",
        cache: "no-store",
        headers: { "Content-Type": "text/plain" },
        body: packed
      }).then(function (r) {
        if (!r.ok) return send(i + 1);
        markSent(digest, via);
        return true;
      }, function () {
        return send(i + 1);
      });
    }
    return send(0);
  }

  function matchingSlot(spec, digests) {
    var i;
    var pc = slotHash(spec, "pc");
    var phone = slotHash(spec, "phone");
    for (i = 0; i < digests.length; i++) {
      if (pc && digests[i] === pc) return "pc";
      if (phone && digests[i] === phone) return "phone";
    }
    return "";
  }

  function fillClaim(claim) {
    var other = document.querySelector("[name=from_other]");
    if (other && String(other.value || "").trim()) return;
    document.querySelectorAll('input[name="from"]').forEach(function (el) {
      if (el.type === "hidden") return;
      if (document.activeElement === el) {
        var cur = String(el.value || "").trim().toUpperCase();
        if (cur && cur !== claim) return;
      }
      el.value = claim;
    });
    try { localStorage.setItem("commons-from", claim); } catch (e) {}
  }

  function ensureMark() {
    var host = document.getElementById("owner-mark");
    if (host) return host;
    var say = document.getElementById("say");
    if (!say || !say.parentNode) return null;
    host = document.createElement("p");
    host.id = "owner-mark";
    host.className = "law";
    say.parentNode.insertBefore(host, say);
    return host;
  }

  function paintMatch(claim, viaSlot, live) {
    var host = ensureMark();
    if (!host) return;
    host.className = "law";
    host.innerHTML = "this machine matches the " + viaSlot +
      " slot for " + claim +
      " (hashed IP, no login). from= is still a claim. " +
      (live ? "two-slot context display is LIVE; Directive 10 is HALF because richer display context remains OPEN." :
        "two-slot context display is OPEN; Directive 10's richer display context is also OPEN.") +
      ' <a href="' + assetUrl("owner-net.html") + '">owner door</a>';
  }

  function setText(id, text) {
    var el = document.getElementById(id);
    if (el) el.textContent = text;
  }

  function paintPanel(state, spec, digests, matchedSlot, live) {
    var panel = document.getElementById("owner-panel");
    if (!panel) return;
    var digest = (digests && digests[0]) || "";
    var via = deviceVia();
    var blob = digest ? JSON.stringify({ k: "owner-net", sha256: digest, via: via }) : "";
    setText("owner-state", state);
    setText("owner-hash", digest || "(echo unreachable this load)");
    setText("owner-json", blob || "");
    setText("owner-pc-slot", slotHash(spec, "pc") ? "set" : "empty");
    setText("owner-phone-slot", slotHash(spec, "phone") ? "set" : "empty");
    setText("owner-distinct", live ? "yes" : "no");
    setText("owner-via", via);
    var knock = document.getElementById("owner-knock-result");
    if (knock) {
      if (live && matchedSlot) {
        knock.textContent = "two-slot context display is LIVE. this machine is the " + matchedSlot +
          " slot. Directive 10 is HALF; this is display context only.";
      } else if (matchedSlot) {
        knock.textContent = "this machine matches the " + matchedSlot + " slot. still OPEN: need the other machine on a different public IP.";
      } else if (!digest) {
        knock.textContent = "echo missed. try again from the phone on cell or the PC at home.";
      } else {
        knock.textContent = "OPEN. empty hashes is not live. this machine is not in either slot yet.";
      }
    }
    var copyBtn = document.getElementById("owner-copy");
    if (copyBtn) copyBtn.disabled = !blob;
  }

  function boot() {
    var panel = document.getElementById("owner-panel");
    loadSpec().then(function (spec) {
      if (!spec || spec.algo !== "sha256") {
        spec = {
          claim: ONLY_CLAIM,
          algo: "sha256",
          pepper: FALLBACK_PEPPER,
          slots: { pc: null, phone: null },
          hashes: []
        };
      }
      if (spec.claim && spec.claim !== ONLY_CLAIM) {
        if (panel) paintPanel("bad spec", spec, [], "", false);
        return;
      }
      return collectDigests(spec).then(function (digests) {
        var live = slotsDistinct(spec);
        var matchedSlot = matchingSlot(spec, digests);
        var matched = !!matchedSlot;
        if (matched) {
          fillClaim(ONLY_CLAIM);
          if (!panel) paintMatch(ONLY_CLAIM, matchedSlot || deviceVia(), live);
        }
        if ((matched || rememberedBryce()) && digests.length) {
          digests.forEach(function (h) { publishDigest(h); });
        }
        if (panel) {
          var state = live ? "TWO-SLOT CONTEXT LIVE — DISPLAY ONLY" :
            (matchedSlot ? "OPEN — this machine matches " + matchedSlot :
              "OPEN — waiting for distinct pc and phone context slots");
          paintPanel(state, spec, digests, matchedSlot, live);
        }
      });
    }).catch(function () {
      if (panel) paintPanel("owner door missed this load", null, [], "", false);
    });
  }

  function bindCopy() {
    var btn = document.getElementById("owner-copy");
    if (!btn) return;
    btn.addEventListener("click", function () {
      var el = document.getElementById("owner-json");
      var text = el ? String(el.textContent || "").trim() : "";
      if (!text || !navigator.clipboard) return;
      navigator.clipboard.writeText(text).catch(function () {});
    });
  }

  function bindKnock() {
    var btn = document.getElementById("owner-knock");
    if (!btn) return;
    btn.addEventListener("click", function () {
      setText("owner-knock-result", "knocking…");
      boot();
    });
  }

  function start() {
    bindCopy();
    bindKnock();
    boot();
  }

  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", start);
  } else {
    start();
  }
})();

File: owner_net.py
#!/usr/bin/env python3
"""Persist phone+PC owner-net digests into owner.json. No raw IPs.

Cite: BRYCE-1787134106972-vr8fo8. Do not remint.
Law: admin-no-verification-loop-20260819-01. Do not remint.

Same-NAT is not Dir 10. Phone on cell and PC at home are two public IPs.
This writer keeps at most two slots (pc, phone). It will not store a phone
digest that equals the pc digest. A filled slot is not overwritten, except
a phone slot that still equals pc when a different digest arrives.

Does not print IPs. Does not touch board_ingest.py.
"""
from __future__ import annotations

import json
import os
import re
import sys
import urllib.error
import urllib.request

ROOT = os.path.dirname(os.path.abspath(__file__))
SPEC_PATH = os.path.join(ROOT, "owner.json")
NET_TOPIC = "woahwhattheheck-commons-owner-net"
NET_HOSTS = (
    "https://ntfy.sh",
    "https://ntfy.envs.net",
    "https://ntfy.adminforge.de",
    "https://ntfy.mzte.de",
)
HASH_RE = re.compile(r"^[0-9a-f]{64}$")
VIAS = ("pc", "phone")
USER_AGENT = "commons-owner-net"


def slot_hash(slot) -> str:
    if isinstance(slot, dict):
        h = str(slot.get("sha256") or "").strip().lower()
        return h if HASH_RE.match(h) else ""
    if isinstance(slot, str):
        h = slot.strip().lower()
        return h if HASH_RE.match(h) else ""
    return ""


def load_spec(path: str = SPEC_PATH) -> dict:
    if not os.path.isfile(path):
        return {
            "claim": "BRYCE",
            "algo": "sha256",
            "pepper": "commons-owner-v1",
            "slots": {"pc": None, "phone": None},
            "hashes": [],
        }
    with open(path, encoding="utf-8") as f:
        spec = json.load(f)
    if not isinstance(spec, dict):
        raise SystemExit("owner.json is not an object")
    slots = spec.get("slots")
    if not isinstance(slots, dict):
        spec["slots"] = {"pc": None, "phone": None}
    else:
        spec["slots"] = {
            "pc": slots.get("pc"),
            "phone": slots.get("phone"),
        }
    return spec


def distinct_live(spec: dict) -> bool:
    pc = slot_hash((spec.get("slots") or {}).get("pc"))
    phone = slot_hash((spec.get("slots") or {}).get("phone"))
    return bool(pc and phone and pc != phone)


def refresh_hashes(spec: dict) -> None:
    hashes = []
    seen = set()
    slots = spec.get("slots") or {}
    for via in VIAS:
        h = slot_hash(slots.get(via))
        if not h or h in seen:
            continue
        seen.add(h)
        hashes.append({"sha256": h, "via": via})
    spec["hashes"] = hashes
    spec["claim"] = "BRYCE"
    spec["algo"] = "sha256"


def apply_sighting(spec: dict, digest: str, via: str) -> bool:
    """TOFU two-slot write. Returns True if owner.json would change."""
    h = str(digest or "").strip().lower()
    v = str(via or "").strip().lower()
    if not HASH_RE.match(h) or v not in VIAS:
        return False
    slots = spec.setdefault("slots", {"pc": None, "phone": None})
    pc = slot_hash(slots.get("pc"))
    phone = slot_hash(slots.get("phone"))
    if v == "pc":
        if pc:
            return False
        slots["pc"] = {"sha256": h}
        refresh_hashes(spec)
        return True
    # phone: never enroll the pc digest as the phone slot.
    if h == pc and pc:
        return False
    if not phone:
        slots["phone"] = {"sha256": h}
        refresh_hashes(spec)
        return True
    # phone was filled with the wifi (same as pc). Replace when cell appears.
    if phone == pc and h != pc:
        slots["phone"] = {"sha256": h}
        refresh_hashes(spec)
        return True
    return False


def parse_payload(raw: str) -> tuple[str, str]:
    try:
        obj = json.loads(raw)
    except json.JSONDecodeError:
        return "", ""
    if not isinstance(obj, dict) or obj.get("k") != "owner-net":
        return "", ""
    return slot_hash(obj), str(obj.get("via") or "").strip().lower()


def poll_host(host: str) -> list[tuple[str, str]]:
    url = "%s/%s/json?poll=1&since=12h" % (host, NET_TOPIC)
    req = urllib.request.Request(
        url, headers={"Accept": "application/x-ndjson", "User-Agent": USER_AGENT}
    )
    try:
        with urllib.request.urlopen(req, timeout=20) as resp:
            raw = resp.read().decode("utf-8", "replace")
    except (urllib.error.URLError, TimeoutError, OSError):
        return []
    out = []
    for line in raw.splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            ev = json.loads(line)
        except json.JSONDecodeError:
            continue
        digest, via = parse_payload(str(ev.get("message") or ""))
        if digest and via:
            out.append((digest, via))
    return out


def collect_sightings() -> list[tuple[str, str]]:
    for host in NET_HOSTS:
        rows = poll_host(host)
        if rows:
            return rows
    return []


def persist(spec: dict, sightings: list[tuple[str, str]]) -> int:
    n = 0
    for digest, via in sightings:
        if apply_sighting(spec, digest, via):
            n += 1
    return n


IPV4_BLOB = re.compile(r"\b(?:\d{1,3}\.){3}\d{1,3}\b")
IPV6_BLOB = re.compile(r"\b[0-9a-fA-F:]+:[0-9a-fA-F:]+\b")


def refuse_raw_ips(blob: str) -> None:
    if IPV4_BLOB.search(blob) or IPV6_BLOB.search(blob):
        raise SystemExit("owner.json would contain a raw IP — refusing to write")


def write_spec(spec: dict, path: str = SPEC_PATH) -> None:
    refresh_hashes(spec)
    blob = json.dumps(spec, indent=2) + "\n"
    refuse_raw_ips(blob)
    with open(path, "w", encoding="utf-8") as f:
        f.write(blob)


def main(argv: list[str] | None = None) -> int:
    argv = list(sys.argv[1:] if argv is None else argv)
    spec = load_spec()
    sightings = collect_sightings()
    changed = persist(spec, sightings)
    if changed:
        write_spec(spec)
    live = "LIVE" if distinct_live(spec) else "OPEN"
    print("owner-net slots pc=%s phone=%s distinct=%s wrote=%d" % (
        "yes" if slot_hash((spec.get("slots") or {}).get("pc")) else "no",
        "yes" if slot_hash((spec.get("slots") or {}).get("phone")) else "no",
        live,
        changed,
    ))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
